python/lsst/daf/persistence/butler.py

```diff
@@ -822,11 +828,50 @@ def dataRef(self, datasetType, level=None, dataId={}, **rest):
         return ButlerDataRef(subset, subset.cache[0])
 
     def _read(self, location):
-        self.log.debug("Starting read from %s", location)
-        results = location.repository.read(location)
-        if len(results) == 1:
-            results = results[0]
-        self.log.debug("Ending read from %s", location)
+        """Unpersist an object using data inside a butlerLocation object.
+
+        A weakref to loaded objects is cached here. If the object specified by the butlerLocaiton has been
+        loaded before and still exists then the object will not be re-read. A ref to the already-existing
+        object will be reutrned instead.
+
+        Parameters
+        ----------
+        location - ButlerLocation
+            A butlerLocation instance populated with data needed to read the object.
+
+        Returns
+        -------
+        object - an instance of the object specified by the butlerLoction.
+            The object specified by the butlerLocation will either be loaded from persistent storage or will
+            be fetched from the object cache (if it has already been read before).
+        """
+        def hasher(butlerLocation):
+            """Hash a butler location for use as a key in the object cache."""
+            try: # is the dataId a daf_persistence.Policy?
+                usedDataId = butlerLocation.usedDataId.data # if so, fetch the dict
+            except AttributeError: # if it's not,
+                usedDataId = butlerLocation.usedDataId # assume it's a raw dict.
+            return hash((butlerLocation.pythonType, butlerLocation.cppType, butlerLocation.storageName,
+                         id(butlerLocation.mapper), id(butlerLocation.storage),
+                         tuple(butlerLocation.locationList), json.dumps(usedDataId),
+                         butlerLocation.datasetType))
```

Comment:
It might be worthwhile to move this to a `ButlerLocation` method so it can be used elsewhere.  I think we may want to cache objects in `put` as well, for instance (I can definitely think of use cases for that, though I haven't thought hard about any problems it might create).